### PR TITLE
Add torch.library.opcheck

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124496

This PR:
- exposes torch.testing._internal.optests.opcheck as
  torch.library.opcheck
- Adds support for CustomOpDef (aka functions decorated with
  torch.library.custom_op) to opcheck.

Test Plan:
- Updated tests
- We validated opcheck's design internally.

### CODE REVIEW

**Review by williamwen42 — APPROVED** [review]

File: docs/source/library.rst [hunk 1]
```diff
@@ -4,13 +4,23 @@ torch.library
...
+custom operators, and extending operators defined with PyTorch's C++ operator
 registration APIs (e.g. aten operators).
 
 For a detailed guide on effectively using these APIs, please see
 `this gdoc <https://docs.google.com/document/d/1W--T6wz8IY8fOI0Vm8BF44PdBgs283QvpelJZWieQWQ/edit>`_
 
+Testing custom ops
+------------------
+
+Use :func:`torch.library.opcheck` to test custom ops for incorrect usage of the
```
Comment (albanD) [inline docs/source/library.rst:17]: Why not `torch.testing.custom_op.opcheck` ?

Comment (zou3519) [reply docs/source/library.rst:17]: It's probably too many namespaces for people to keep track of? The analogy is torch.autograd.gradcheck -- would we have wanted to put it into torch.testing.autograd.gradcheck?

Comment (albanD) [reply docs/source/library.rst:17]: torch.testing didn't exist when gradcheck was added?
I'm happy to open the discussion to move gradcheck there if you think that we should do a larger effort.

Comment (zou3519) [reply docs/source/library.rst:17]: Resolved offline: we'll go with torch.library.opcheck for now, but we're happy to put it into torch.testing too

**Pull request closed without merging**